### DISCUSSION TITLE
Incorrect results in WEC-Sim for a slender horizontal cylinder

### DISCUSSION THREAD

**Discussion by ajay-menon-iitkgp** [capytaine/capytaine#444] [Q&A] [answered]
Hello, Community! I've recently started using Capytaine to provide the hydroData for my WEC-Sim simulations.

### Context
I am simulating a **slender horizontal cylinder** (length 1.8 m, radius 0.1 m) in WEC-Sim, and have used the `mesh_horizontal_cylinder()` Capytaine function to build the body. Other details are: infinite water depth and 0-degree wave heading. All relevant files have been attached below:
[Capytaine Discussion - Files.zip](https://github.com/capytaine/capytaine/files/13711943/Capytaine.Discussion.-.Files.zip)

The `1-segment.py` Capytaine file and its output are in the `hydroData` directory.

### Problem Description
WEC-Sim yields incorrect results for the slender cylinder (video included below). The steps I undertake are:
1. Run the `1-segment.py` file and export output to my WEC-Sim $CASE hydroData directory.
2. Run `bemio.m` in this hydroData directory.
3. Prepare WEC-Sim input file and Simulink model.
4. Execute `wecSim` in the main simulation directory.

### WEC-Sim Explorer Output
Result sped up by **factor of 4** (original sim time = 200 seconds, ramp time = 100 seconds)
https://github.com/capytaine/capytaine/assets/77717855/ba8d9100-28b1-4dd8-8f8a-7cd03231a02d

The body pitches uncontrollably, which leads me to suspect that there is an issue with the damping/added mass (please correct me if this is an incorrect conclusion).

### What I Have Tried
1. I assumed that the weird results above were caused by improper mesh resolution. But increasing the # of panels up to 3,600 does not improve the results.
2. Likewise, increasing the number of frequency components to 100, within 0.1 - 10 rad/s does not improve the results.
3. Changing the cylinder's radius and length by a small factor to see if the issue is caused by its **slender-ness** does not improve the results.

### Capytaine Code
To quickly refer to, my Capytaine code is as follows (it is also included in the hydroData directory):
```
# Import statements
import numpy as np
import xarray as xr
import capytaine as cpt

# Legacy Capytaine import statements
from capytaine.io.legacy import export_hydrostatics
from capytaine.io.legacy import write_dataset_as_tecplot_files
from capytaine.io.xarray import separate_complex_values
from capytaine.io.xarray import merge_complex_values

# Import logging module and set to 'debug' level
import logging
logging.basicConfig(level=logging.DEBUG)

# Define mesh for a body labelled 'segment_1'
segment_1 = cpt.FloatingBody(
                      cpt.mesh_horizontal_cylinder(
                          length=1.8, radius=0.1,
                          center=(0.0, 0.0, 0.0),
                          resolution=(10,10,50)),
                          name="segment_1")
segment_1.center_of_mass = np.array([0.0, 0.0, 0.0])
segment_1.rotation_center = np.array([0.0, 0.0, 0.0])
segment_1.add_all_rigid_body_dofs()
segment_1.compute_hydrostatics()
segment_1.inertia_matrix = segment_1.compute_rigid_body_inertia()
segment_1.hydrostatic_stiffness = segment_1.compute_hydrostatic_stiffness()
segment_1.keep_immersed_part()

# Visualize mesh using the 'vtk' python library
segment_1.show()

# Define range of frequencies
omega_range = np.linspace(0.01,10,100)

# Define the radiation and diffraction problems for the floating body
problems = [cpt.RadiationProblem(body=segment_1, radiating_dof=dof, omega=omega)
            for dof in segment_1.dofs
            for omega in omega_range]
problems += [cpt.DiffractionProblem(body=segment_1, wave_direction=0.0, omega=omega)
             for omega in omega_range]

# Solve the Capytaine problem
solver = cpt.BEMSolver()
results = solver.solve_all(problems)
dataset = cpt.assemble_dataset(results)

# Export output data
separate_complex_values(dataset).to_netcdf("D:/Capytaine/Trial files/segment-new/results/dataset.nc",
                  encoding={'radiating_dof': {'dtype': 'U'},
                            'influenced_dof': {'dtype': 'U'}})
dataset = merge_complex_values(xr.open_dataset("D:/Capytaine/Trial files/segment-new/results/dataset.nc"))
export_hydrostatics("D:/Capytaine/Trial files/segment-new/results",segment_1)
write_dataset_as_tecplot_files("D:/Capytaine/Trial files/segment-new/results",dataset)
```

### Files and Screenshots
These are the BEMIO outputs:
![Added_Mass](https://github.com/capytaine/capytaine/assets/77717855/03d66f74-4361-489d-8e1f-2a45840891af)
![Excitation_IRFs](https://github.com/capytaine/capytaine/assets/77717855/02dab4a6-5b6b-470d-88bd-f02cd6909201)
![Excitation_Magnitude](https://github.com/capytaine/capytaine/assets/77717855/f746bad3-b92b-4366-ac1b-47f2993521a3)
![Excitation_Phase](https://github.com/capytaine/capytaine/assets/77717855/5639cbef-0003-4d44-ad51-b5a02b209e38)
![Radiation_Damping](https://github.com/capytaine/capytaine/assets/77717855/57384d7f-5505-4517-a268-7ab0c1eebb3d)
![Radiation_IRFs](https://github.com/capytaine/capytaine/assets/77717855/d76a554c-c066-4484-9b4c-51fe071ec41e)
**End of screenshots**

I'm sure this is a basic problem, so I must be making an obvious error somewhere. Thank you for taking a look at it and I'd appreciate any comments/feedback.

Cheers!

**Comment by mancellin** [comment]
Thank you for the detailed report.
From a quick look, I don't see any obvious mistake in the code.
I'll take a deeper look later.
In the mean time, I'm pinging @dav-og @salhus which are more familiar than me with WEC-SIM.

**Reply by ajay-menon-iitkgp** [reply]
Much appreciated, Matthieu. Cheers!

**Reply by salhus** [reply, marked as answer]
@ajay-menon-iitkgp 
Hi,

The body is $1.8 m$ long, while the wave height is $2.5 m$. The seemingly uncontrolled pitching doesn't happen for smaller wave heights. The wave height of $2.5 m$ violates the linear conditions, and the hydrostatic coefficients won't be accurate anymore. I would suggest using the non-linear hydrostatics option should you want to forgo the hydrostatic coefficient matrix.

Also, the BEM should be run for even higher frequencies, so that the IRFs don't have that chatter because the frequency-response/hydrodynamic coefficients were truncated before convergence. Note, I am pointing out the range of the BEM frequencies not their resolution.

Cheers,
Sal

**Reply by ajay-menon-iitkgp** [reply]
Thanks for taking a look at the problem, @salhus ! It's a long reply, so I've broken it down by your suggestions and how I've addressed them.

#### Non-linear hydrodynamics
I forgot that I'd have to specify non-linear hydrodynamics because of how large the body is compared to the wave height. I've referred to the WEC-Sim documentation and have updated `wecSimInputFile.m` as follows:

1. `body(1).nonlinearHydro = 2` to include all non-linear effects,
2. Viscous drag - approximate quadratic drag coefficients and area,
3. Refined the body mesh as recommended in the WEC-Sim tutorial.

Can you confirm that these are the only changes to be made for the inclusion of non-linear effects (screenshot of code below)?
![Screenshot 2023-12-20 172518](https://github.com/capytaine/capytaine/assets/77717855/d42a5b53-1115-4eb4-bc83-dd6f87d707c3)

#### Increased range of frequencies in Capytaine 
As you suggested, I ran Capytaine for 30 seconds at 1000 frequency components (is that too much?) and noticed that the BEMIO plots were reaching steadier values than before.
![Added_Mass](https://github.com/capytaine/capytaine/assets/77717855/55a676a9-0e30-40b4-bdf5-e4f2f16a8c3a)
![Excitation_IRFs](https://github.com/capytaine/capytaine/assets/77717855/31529b15-d89c-484c-97fe-af5f30230cd9)
![Radiation_Damping](https://github.com/capytaine/capytaine/assets/77717855/fa18cae3-8510-4e9c-86ca-1222cfc780f4)
![Radiation_IRFs](https://github.com/capytaine/capytaine/assets/77717855/77661a7d-4d29-4537-ba52-f8a2044bb274)

I noticed odd spikes from irregular frequencies in the damping and added mass plots. **Question:** do I need to manually smooth these out or do you think it might not affect the simulation too much?

#### Improved mesh
On another issue I'd raised (WEC-Sim/WEC-Sim#1179), you mentioned that I need to improve the mesh quality. I exported the Capytaine mesh as a .stl file for the WEC-Sim input. The mesh is now more orderly and refined. However, when I run WEC-Sim, I think there must be a confusion with the dimensions because Mechanics Explorer shows a **1.8mm** cylinder, not **1.8m**. The simulation executes, though, without returning any errors. Is there any way to specify dimensions in either Capytaine or WEC-Sim?

While WEC-Sim usually uses the .stl file only to visualize, the non-linear tutorial mentions that a refined mesh is required to compute the non-linear pressure forces. So, I now export a cylinder from Capytaine with particulars of 1800 (length) and 100 (radius), assuming that WEC-Sim is reading it in mm. But, I am running BEMIO on the original results from Capytaine (1.8, 0.1). While the cylinder is now correctly visualized in WEC-Sim, the simulation crashes prematurely.

#### Error after implementing changes
The error message displayed is:
```
Error using wecSim
Derivative of state 'cylinder.Floating_3DOF.CONSTRAINT_Planar_Joint.Px.v' in block 'cylinder/Floating (3DOF)/CONSTRAINT
(Planar Joint)' at time 0.034999999999999996 is not finite. The simulation will be stopped. There may be a singularity in the
solution.  If not, try reducing the step size (either by reducing the fixed step size or by tightening the error tolerances)
```

From other issues and the discussion page on WEC-Sim, other users were able to resolve the issue by changing the solver to variable time-step (ode45) or reducing the time step size (to 0.01 or lower). However, neither fix worked in this case. An interesting observation is the simulation does not crash when I use the 1.8-specified cylinder.stl file, but only for the larger 1800-specified cylinder.stl file.

#### Reproducing the error/issue
My new WEC-Sim and Capytaine files are below:
[Capytaine Disucssion - Files v2.zip](https://github.com/capytaine/capytaine/files/13726121/Capytaine.Disucssion.-.Files.v2.zip)

In the hydroData folder, I've included my `export_mesh.py` script to export the Capytaine mesh to a .stl format.

Thank you!

**Reply by ajay-menon-iitkgp** [reply]
Hi, @salhus. I've just compared the BEM results between Capytaine and Nemoh after incorporating your feedback, and both match perfectly! As you pointed out, I should have run the BEM code for a larger $\omega$ range. I ran a few different cases, and increasing the upper limit to $30$ $rad/s$ produces stable results. The added mass reaches a constant value while the damping and IRFs tend to $0$.

I realize this is not an issue with Capytaine, but rather how I'm setting up WEC-Sim. Do you recommend I transfer our discussion to issue WEC-Sim/WEC-Sim#1179?

One question (relevant to Capytaine) that I still have is regarding the mesh dimensions when I import the .stl file into WEC-Sim. I notice that neither tool allows me to specify the dimensions. So, assuming meter is the default unit, why does WEC-Sim import the mesh in millimeters? Is there a work around for this or a better way to proceed? @mancellin

**Reply by mancellin** [reply]
> One question (relevant to Capytaine) that I still have is regarding the mesh dimensions when I import the .stl file into WEC-Sim. I notice that neither tool allows me to specify the dimensions. So, assuming meter is the default unit, why does WEC-Sim import the mesh in millimeters? Is there a work around for this or a better way to proceed? @mancellin

Yes, Capytaine assumes the data in the STL file are in meters (see also https://github.com/capytaine/capytaine/issues/74).

The only reference I found about WEC-SIM is [this line in its source code](https://github.com/WEC-Sim/WEC-Sim/blob/df50a039e3af5d4a01ca2f2e436fc15400277b4e/source/objects/bodyClass.m#L503) according to which WEC-SIM also uses meters.

If necessary, you can rescale the data before saving them to a file, something like:
```python
from capytaine.io.mesh_writers import write_STL
write_STL("mesh.stl", body.mesh.vertices*1000, body.mesh.faces)  # STL file in millimeters
```

**Reply by ajay-menon-iitkgp** [reply]
Right, so it seems like I will need to rescale the data before exporting to an STl file. I'll check #74 and check where the mix-up happens.
Thank you very much, Matthieu!

**Reply by salhus** [reply]
@ajay-menon-iitkgp 
I saw this thread after typing my comment for the WEC-Sim thread. And yes, this confirms the mm vs. m issue..  
I think it has something to do with how the stl binaries are being read by Simscape.  However, I think @mancellin recommendation of using the capytaine IO might be a quick open-source solution to this.

**Reply by salhus** [reply]
@ajay-menon-iitkgp 
re: irregular frequencies,
You can usually be mostly okay, as long as the irregular frequencies don't make the IRF too noisy.  You could use BEM Rosetta to clean up the irregular frequencies should you want to. If you avoid input waves around irregular frequencies, you should be fine for most cases.
Cheers,
sal

**Reply by ajay-menon-iitkgp** [reply]
Happy New Year!
@salhus The Capytaine solution seems to resolve the issue with the units. Thank you for your comments on the irregular frequencies. I'll explore cleaning the IRF in BEM Rosetta and check whether reducing the frequency resolution might resolve it.

Appreciate both of your responses and for helping me resolve this issue. Good day!

**Reply by salhus** [reply]
@ajay-menon-iitkgp 
Note, BEM Rosetta will clean the hydrodynamic coefficients (frequency domain), as a consequence of which the IRFs get cleaned up.
